addons/netfox.extras/network-simulator.gd

```diff
@@ -0,0 +1,188 @@
+@export_range(0, 100) var packet_loss_percent: float = 0.0
+
+var enet_peer: ENetMultiplayerPeer = ENetMultiplayerPeer.new()
+
+# UDP proxy
+var proxy_thread: Thread
+var udp_proxy_server: PacketPeerUDP
+var udp_proxy_port: int = 9998
+var rng_packet_loss: RandomNumberGenerator = RandomNumberGenerator.new()
+
+# Connection tracking
+var client_queues: Dictionary = {} # port to data stream  - Array[QueueEntry]
+var client_peers: Dictionary = {} # port to PacketPeerUDP
+var server_to_client_queue: Array[QueueEntry] = []
+
+class QueueEntry:
+	var packet_data: PackedByteArray
+	var queued_at: int
+	var target_port: int
+	
+	func _init(packet: PackedByteArray, timestamp: int, port: int = -1) -> void:
+		self.packet_data = packet
+		self.queued_at = timestamp
+		self.target_port = port
+
+func _ready() -> void:
+	await get_tree().process_frame
+	udp_proxy_port = server_port + 1
+	
+	if OS.has_feature("editor"):
+		var status = try_and_host()
+		if status == Error.ERR_CANT_CREATE:
+			try_and_join()
+		
+		multiplayer.multiplayer_peer = enet_peer
+
+func try_and_host() -> Error:
+	var status = enet_peer.create_server(server_port)
+	if status == OK:
+		if is_proxy_required():
+			start_udp_proxy()
+		server_created.emit()
+	return status
+
+func try_and_join() -> Error:
+	var connect_port = server_port
+	if is_proxy_required():
+		connect_port = udp_proxy_port
+	var status = enet_peer.create_client(hostname, connect_port)
+	if status == OK:
+		client_connected.emit()
+	return status
+
+# Starts a UDP proxy server to simulate network conditions
+# This will listen on udp_proxy_port and forward packets to the server_port
+# Runs on its own thread to avoid blocking the main thread
+func start_udp_proxy() -> void:
+	proxy_thread = Thread.new()
+	udp_proxy_server = PacketPeerUDP.new()
+	
+	var bind_status = udp_proxy_server.bind(udp_proxy_port, hostname)
+	if bind_status != OK:
+		push_error("Failed to bind UDP proxy port: ", bind_status)
+		return
+	
+	proxy_thread.start(process_packets)
+
+
+func process_packets() -> void:
+	while true:
+		var current_time: int = Time.get_ticks_msec()
+		var send_threshold: int = current_time - latency_ms
+		
+		_read_client_to_server_packets(current_time)
+		_process_client_packets(send_threshold)
+
+		if not client_queues.is_empty():
+			_read_server_to_client_packets(current_time)
+			_process_server_to_client_queue(send_threshold)
+
+func _read_client_to_server_packets(current_time: int) -> void:
+	while udp_proxy_server.get_available_packet_count() > 0:
+		var packet = udp_proxy_server.get_packet()
+		var err = udp_proxy_server.get_packet_error()
+		
+		if err != OK:
+			push_error("UDP proxy incoming packet error: ", err)
+			continue
+		
+		var from_port = udp_proxy_server.get_packet_port()
+		_register_client_if_new(from_port)
+		
+		var client_queue = client_queues[from_port] as Array[QueueEntry]
+		client_queue.push_back(QueueEntry.new(packet, current_time))
+
+func _register_client_if_new(port: int) -> void:
```

Comment:
Would it make sense to use a single queue? That way we don't need a nested loop of client queues and then each entry in the queue. Unless there's some secret feature planned that I don't know of 😀